### ISSUE TITLE
Refactor tail-sampling processor

### ISSUE THREAD

**Issue by jpkrohling** [open-telemetry/opentelemetry-collector-contrib#31580]
### Component(s)

processor/tailsampling

### Is your feature request related to a problem? Please describe.

The tail-sampling processor has been there for a few years now, and despite being relatively popular, it hasn't changed much over the years besides adding new policies. There are a few things I always wanted to do but didn't have the time or motivation, and it would be a good idea to list them here in case we have volunteers to work on them.

### Describe the solution you'd like

- [ ] Move to OTel metrics instead of OpenCensus metrics
- [ ] Refactor policies
- [ ] Decision cache
- [ ] Experiment and validate the scheduling mechanism

**Comment by jpkrohling** [comment]
cc @kentquirk

**Comment by kentquirk** [comment]
Yes, AND, I also think that the binary keep/drop structure is a problem because it's not adequately composable if the probability of the decision matters to you. [This proposal](https://github.com/open-telemetry/oteps/pull/250/files) is related but I'm not sure yet if it actually applies.

**Comment by jiekun** [comment]
I'm interested in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31584  if we could have more info / context / expectation :P

**Comment by jpkrohling** [comment]
I think this is also relevant for the discussion: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/25122

**Comment by jpkrohling** [comment]
Related: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31498

**Comment by jpkrohling** [comment]
Unassigning myself — I don't have the bandwidth to pursue this anymore.
